Turn user.metadata into accessor class - user.get_user_data(domain)

esoergel opened the pull request:
Ready for review, though I'm keeping it in draft until I'm satisfied with testing

## Product Description
No meaningful product changes

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-3706

Big picture, this is a step in the direction of giving web users access to user data.  They sorta have it already, but the problem is conceptually, `user_data` is project-specific.  This means it belongs on the `DomainMembership` model, so web users on multiple domains can have user data dicts scoped to each domain.

To enable moving `user.user_data` to the `DomainMembership` model, this PR moves all access to that field to a domain-scoped accessor.  This way, we can (in the next ticket in this series) change the underlying data storage details and only  update the definition of `CouchUser.get_user_data` without affecting how the callers interact with it.

Since this is sort of a dictionary, but with some special additions on top, I made an accessor class that quacks like a dictionary, but guards against invalid use and factors in some multiple places that data comes from.


## Feature Flag


## Safety Assurance

### Safety story

I tested out a lot of this locally, and added test coverage for how it should work.

### Automated test coverage


### QA Plan

Planning on putting this in QA on staging


### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
move profile updated logic to caller

esoergel committed:
Extract metadata tests

esoergel committed:
Simplify profile getter

and move error handling into caller

esoergel committed:
Add new class to manage user data

esoergel committed:
Split up giant test for clarity

otherwise you have to keep track of the ever-mutating user data dict

esoergel committed:
be sure to wipe profile cache

esoergel committed:
Update profile management portion

esoergel committed:
Add wrapper to web users

esoergel committed:
Update some easy references to user.metadata

esoergel committed:
Move metadata stuff to non-db tests

esoergel committed:
Set profile first when updating

esoergel committed:
Ignore profile conflicts if they're the same

I don't love this behavior, but it's consistent with existing code and
I'd prefer not to change it at this time

esoergel committed:
Update user API handlers

Note that the existing code had a bug where it compared `user.metadata`
against `user.user_data` to see if anything.  These two may differ even
if nothing changed.  The updates I made in this commit change that
behavior

esoergel committed:
Use custom exception

esoergel committed:
s/profile/profile_name

esoergel committed:
Test change profile by name, not ID

Profile names are displayed in the download, and users can reassign
profiles by editing that column.  The tests, however, modify profile by
editing the `commcare_profile` custom data field, so they're effectively
testing something else entirely.  Oddly enough, this does work, as long
as you can figure out the profile ID (it's not shown in the UI), and you
upload it as an integer datatype.

This behavior seems wrong, so I'm explicitly disallowing it

esoergel committed:
No need to re lookup the profile name

Now that it comes exclusively _from_ the name

esoergel committed:
remove misleading block of code

remove_unused_custom_fields_from_users_task deletes old custom data
fields regardless of whether they're present, missing, or empty.  When
editing data in the UI, all fields on the schema are saved to the DB,
even when blank.

esoergel committed:
Actually, I don't think these add anything

esoergel committed:
allow profiles to overwrite blank fields

esoergel committed:
update user importer code to use user_data wrapper

snopoke commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -188,35 +190,24 @@ def update_name(self, name):
+            profile_obj = domain_info.profiles_by_name[profile_name]
             data[PROFILE_SLUG] = profile_obj.id
-            for key in profile_obj.fields.keys():
-                self.user.pop_metadata(key)
```
Is it intentional to remove these lines?

esoergel replied:
Yep, that was deliberate.  The old version of this added all stored user data to the passed in `data` a few lines above this, then this loop removed profile-controlled fields from `user.user_data` (not from `data`).  It then called `update_metadata` with `data`, so anything that had been removed from `user.user_data` is going to be added back in anyways.  So it's a question of profile conflicting with data versus data conflicting with profile.

Honestly, I had a lot of trouble teasing out how this method works, and it turns out this [essentially doesn't work](https://dimagi-dev.atlassian.net/browse/USH-2762) in its current form end-to-end, so I feel like there's some leeway to fiddle with the logic.  My aim was largely to turn this into a more straightforward update, and use the test cases to convince myself that the logic is sound.

esoergel committed:
Switch "edit_user" page to use new accessors

esoergel committed:
unset profile properly

esoergel committed:
Improve `update` logic

esoergel committed:
Add "pop" method - this was harder than expected

esoergel committed:
use accessors in custom_data task

esoergel committed:
update user model internal accessors

esoergel committed:
Remove last of update_metadata and pop_metadata

esoergel committed:
Update ES stuff to use user_data

esoergel committed:
Remove last usages of metadata

esoergel committed:
fix some tests

esoergel committed:
switch user.create kwarg back to user_data

Just for consistency.  This also handles it explicitly in the WebUser
and CommCareUser implementations, which will be necessary when we move
the data to the `DomainMembership` model.

esoergel committed:
I missed this reference to `.metadata`

esoergel committed:
commcare_project is included by the accessor

esoergel committed:
Log changes to the raw, stored user data

esoergel committed:
`request.domain` is not always available

esoergel committed:
Remove usages of `user.user_data`

esoergel committed:
Push remove_unrecognized logic into accessor

esoergel committed:
Move helper function in-line

MartinRiese commented on corehq/apps/users/user_data.py:
```diff
@@ -0,0 +1,136 @@
+            COMMCARE_PROJECT: self.domain,
+        }
+
+    def to_dict(self):
```
I find this function name a bit confusing. It does not convert the UserData to a different format it just combines the values that are held separately internally. I cannot think of a better one of the top of my head though.

esoergel replied:
My thought process there was that `UserData` is meant to be interacted with like a dictionary, though it's the combination of a few different sources.  That is, you can generally do
```python
user_data = UserData(...)
user_data["name"] = "Ethan"
print(user_data["commcare_location_id"])
```
and so forth, but we do need guards around what can and can't be set directly, since some fields care controlled by the profile.  I added the `to_dict()` method for places that just need a normal python dictionary, similar to how `JsonObject` and `Document` subclasses have `to_json()` methods.

MartinRiese replied:
I suppose `to_dict` is fine

MartinRiese commented on corehq/apps/users/user_data.py:
```diff
@@ -0,0 +1,136 @@
+        # should be consistent
+        original = self.to_dict()
+        for k in data:
+            self._local_to_user.pop(k, None)
```
If you are not using the returned value `del` [might be faster](https://stackoverflow.com/questions/24547223/is-del-or-pop-preferred-when-removing-elements-from-dicts)

esoergel replied:
I was not aware of that, though it requires an extra line of code

MartinRiese reviewed (commented):
Just a quick look. Will keep going tomorrow.

MartinRiese commented on corehq/apps/user_importer/tests/test_importer.py:
```diff
@@ -164,14 +164,18 @@ def _get_spec(self, delete_keys=None, **kwargs):
         spec.update(kwargs)
         return spec
 
+    def assert_user_data_item(self, key, expected):
```
👍

MartinRiese commented on corehq/apps/users/tasks.py:
```diff
@@ -295,10 +295,20 @@ def reset_demo_user_restore_task(commcare_user_id, domain):
     )
-    remove_unused_custom_fields_from_users(domain)
+    fields_definition = CustomDataFieldsDefinition.get(domain, CUSTOM_USER_DATA_FIELD_TYPE)
+    assert fields_definition, 'remove_unused_custom_fields_from_users called without a valid definition'
```
is this message still correct? Does the PR not remove `remove_unused_custom_fields_from_users`. Might be misleading when debugging.

esoergel committed:
translate

esoergel committed:
correct fn name

esoergel committed:
lint

esoergel committed:
lint

esoergel committed:
Remove some old code no longer used

esoergel committed:
Set profile separately from everything else

Add stub save() method

stephherbers committed:
Merge branch 'master' into es/metadata

esoergel committed:
fix bad merge

esoergel committed:
update test

esoergel committed:
remove unused metadata arg

esoergel committed:
move get_custom_data to CouchUser

esoergel commented:
QA ticket: https://dimagi-dev.atlassian.net/browse/QA-5751

snopoke commented:
@esoergel how did you resolve the consistency issue with saving a user?

Also, it would be good to update the README in the `users` app.

esoergel commented:
@snopoke ah thanks I'll do that, that wasn't on my radar.  This PR doesn't affect the saving yet, that's in the followup PR:
https://github.com/dimagi/commcare-hq/pull/33688 (still a WIP).  That PR calls `save()` on the SQL model inside `user.save()`.  I attached the django user to get cascading deletion, but that broke a bunch of `SimpleTestCase` tests - working on figuring that out now.
